### ISSUE TITLE
[Feature] Add `jprint(1)` tool to output valid JSON in canonical format

### ISSUE THREAD

**Issue by lcn2** [xexyl/jparse#5]
### Is there an existing issue for this?

- [X] I have searched for existing issues and did not find anything like this

### Describe the feature

We recommend adding a `jprint(1)` tool that will output JSON that has been parsed and found to be valid, in canonical format.

For the canonical output format model, we recommend that [bin/jprint-wrapper.sh](https://github.com/ioccc-src/temp-test-ioccc/blob/master/bin/jprint-wrapper.sh) prints.

### Relevant images, screenshots or other files

We recommend the following command line for `jprint(1)`:

```
usage: jprint [-h] [-v level] [-V] [--use-jparse] [-S] [-i ext] file.json

    -h		    print help message and exit
    -v level	    set verbosity level (def level: 0)
    -V		    print version string and exit

    --use-jparse    Do nothing: verify this tool came from the jparse toolset

    -S              Reserved for future use in printing JSON semantic strings (def: do not)

    -i  ext         Edit files in-place, saving backups with the specified ext extension, do not write to stdout
                    If a zero-length is given, no backup will be saved.  (def: do not edit files)
                    If file.json is - (dash) or . (dot) then this option is ignored.

    -c              Compare new contents with the old contents and only update is different
                    The use of -c requires the use of -i ext
                    If file.json is - (dash) or . (dot) then this option is ignored.

    file.json       JSON input file
                    If file.json is - (dash) read from standard input.
                    If file.json is . (dot) do not read any file, perform an internal test and exit accordingly.

Exit codes:
     0         all OK
     1         file is not valid JSON 
     2         cannot read file.json, or file.json does not exist, or file.json is not a file
     3         internal test failed
 >= 10         internal error 
```

### Relevant links

The [bin/jprint-wrapper.sh](https://github.com/ioccc-src/temp-test-ioccc/blob/master/bin/jprint-wrapper.sh) tool is the model for how to print JSON canonically.

### Anything else?

We would not expect another jparse command to be able to undestant `--use-jparse` as a command line option.  Again, the `--use-jparse` command line option does nothing (other than to be parsed as a command line option).

When file.json is - (dash), read from standard input as in:

```sh
    /some/command | jprint -
```

When file.json is . (dot), perform an internal test by parsing an internal (static array in memory) JSON.  If the test of the internal (static array in memory) JSON fails, when `jprint(1)` must exit 3.

The purposes of `jprint --use-jparse` is to "verify" that the `jprint(1)` tool came from this jparse repo toolset.  There are likely to be more than 1 jprint command out there.  Do verify that the jprint tool found on $PATH is the proper jparse tool version we use:

```sh
    jprint --use-jparse .
```

The `-i ext` allow one to update a file with the canonically formatted JSON and save the old contents of the file.  For example:

```sh
    jprint -i .old file.json
```

The previous contents is saved as `file.json.old` and `file.json` is replaced with the canonically formatted JSON contents.  The actions to move files must be performed using the `rename(2)` system call.  The effect of the above command is as if this was performed:

```sh
    jprint file.json > .tmp.file.json
    ln -f file.json file.json.old
    mv -f .tmp.file.json file.json
```

If the `ext` is an empty string, no old file contents is saved:

```sh
    jprint -i "" file.json
```

The `file.json` contents is replaced with the canonical JSON contents canonically formatted JSON contents.  The actions to move files must be performed using the `rename(2)` system call.  The effect of the above command is as if this was performed:

```sh
    jprint file.json > .tmp.file.json
    mv -f .tmp.file.json file.json
```

When the `file.json` is - (dash) or . (dot), the `-i ext` is ignored.

The `-c` option must be used with `-i ext`.  The `file.json` file is **ONLY** touched if the contents of the file is updated.  If the contents of the `file.json` file is not changed, no old file is saved.

This command:

```sh
    jprint -i .old -c file.json
```

has the effect of:

```sh
    jprint file.json > .tmp.file.json
    if cmp -s file.json .tmp.file.json; then
        rm -f .tmp.file.json
    else
        ln -f file.json file.json.old
        mv -f .tmp.file.json file.json
    fi
```

This command:

```sh
    jprint -i "" -c file.json
```

has the effect of:

```sh
    jprint file.json > .tmp.file.json
    if cmp -s file.json .tmp.file.json; then
        rm -f .tmp.file.json
    else
        mv -f .tmp.file.json file.json
    fi
```

**NOTE**: In the above examples, files of the form `.tmp.file.json` represents a temporary file.  One should use the apppriate `mkstemp(3)` call.

The contents of `file.json` should be read into memory.  If the  `file.json` is not a file, or if  `file.json` does not exist, or if there is a read error while reading the file into memory, or if the command runs out of memory, then the `jprint(1)` command must exit 2.

Once the contents of the `file.json` is in memory, then `jparse_json(3)` should be called to parse the contents.  If the contents is **NOT** valid JSON (or the JSON parser function failed), then the `jprint(1)` command must exit 1.

## UPDATE 0

We will save this issue now, and return in the morning to review and if need edit/revise it.  This is just a draft.

**Comment by xexyl** [comment]
Thanks for opening this. Please let me know when you've updated it so I can read it proper.

**Comment by lcn2** [comment]
> Thanks for opening this. Please let me know when you've updated it so I can read it proper.

Going to sleep now.  Will consider this perhaps tomorrow or so as this is not a high priority.

**Comment by xexyl** [comment]
> > Thanks for opening this. Please let me know when you've updated it so I can read it proper.
> 
> Going to sleep now. Will consider this perhaps tomorrow or so as this is not a high priority.

Quite all right .. sleep tight!

**Comment by lcn2** [comment]
# An UPDATE

We are focusing on compelting "bin tools" in the other repo, as well as how to open up a new issue in this repo related to a `jval(1)` tool as discussed in [GH-](https://github.com/ioccc-src/mkiocccentry/issues/523#issuecomment-2323540461).   As such it will be even longer before we address the `-S` concept.  We realize this impacts the [chkentry(1) related issue](https://github.com/ioccc-src/mkiocccentry/issues/940), however given priorities we need to delay addressing this further.

# The BELOW may be edited without warning

We will print this now, and edit things as we think / re-read the text.  This is a "work in progress" as the expression goes.

# Please hold off on asking too many questions about this next section.

Nevertheless to give you a preliminary concept on what are "_JSON semantic strings_".  We do this in the hope that it will **NOT DISTRACT YOU FROM HIGHER PRIORITIES** and instead if you an idea about the general concepts that are being architected.

Consider this valid JSON:

```json
{
    "curds" : 23209,
    "foo" : [
        "bar",
        true,
        42,
        null
    ]
}
```

With `jprint -S`, each JSON value (JSON strong, JSON number, JSON boolean, JSON null) would be printed as a JSON string:

```json
{
    "description of JSON name member, curds" : "description of JSON integer number as a JSON member value of curds",
    "description of JSON name member, foo" : [
        "description of JSON string as part of an array that is a JSON member value of foo",
        "description of JSON boolean as part of an array that is a JSON member value of foo",
        "description of JSON integer number as part of an array that is a JSON member value of foo",
        "description of JSON null as part of an array that is a JSON member value of foo"
    ]
}
```

The _"description of ..."_ would be a specific, program parsable friendly string that describe details about the semantics of the JSON "thing" is represented, including, but not limited to:

- item_type
- sub-types (such as jSON number information, JSON string encoding, etc.)
- min repeat count
- max repeat count
- where applicable a "chk_name" that can refer to a specific check function name to call
- etc.

**NOTE**: The above bullet list is TBD and may be slightly modified as the architecture of the JSON semantics continues to be developed.

The JSON produced by `jprint -S` could be parched / edited to account for variations.   Consider an JSON document that recorded test scores.  One could start with a simple minimum reference example:

```json
[
    100
]
```

With `jprint -S`:

```json
[
    "description of JSON integer number"
]
```

The _"description of JSON integer number"_ would initial indicate a "min repeat count" of 1 and a "max repeat count" of 1.  However if the indent is to have recorded test scores allow for a minimum of 1 and no limit to the number, then the _"description of JSON integer number"_  could easily be patched (modified) to indicate 1+ repeats.

The idea is that a tool such as `chkentry(1)`, when given a JSON document at a JSON semantic document containing JSON semantic strings, would "_walk_" both trees in parallel, checking data in the JSON parse tree and potentially calling check functions as needed.

For JSON such as:

```json
    "email" : "mame@ruby-lang.org",
```

With `jprint -S`:

```json
    "description of JSON name member, email" : "description of JSON integer number as a JSON member value of email",
```

The "description of JSON integer number as a JSON member value of email" patched (modified) to indicate that the value may be a JSON string or a JSON null.  The "where applicable a "chk_name" that can refer to a specific check function name to call" part of the ascription would indicate a "chk_email" function name that when called, would perform further "does it look like a valid Email address or is it JSON null?" check.

## In conclusion

The general concept of having 2 parallel JSON trees, one whose JSON semantics is being checked, and the other that describes the scope of the JSON semantics is being architected to be fairly easy for you to code the general JSON semantics checking function **AND** very flexible to accommodate a wide range of semantics requirements **AND** fairly easy for some programmer who wants to "validate some JSON" beyond a simple "does it match the JSON grammar" check.

The functionally of JSON semantic checking will be an **extremely important** contribution to the overall [jparse repo](https://github.com/xexyl/jparse) and will be something very unique in the world of JSON tools.

The `chkentry(1)` program tried to do some of this, but it was coded too early and the whole `soup/chk.*` mess was an hack to get something working instead of focusing on the more general JSON semantics case.

We intend to finalize the format of the "_JSON semantic strings_", perhaps post-**Great Fork Merge**.  We intend to produce by hand, what `jprint -S` should produce for a reference `.info.json`, a reference `.auth.json`, and a reference `.entry.json` document.

# REQUEST

Try to not let this above "_distract_" to, nor "_alarm_" you.  The architecture is a work in progress, one that is fairly far along but needs some polishing.  It is also something we believe you are very capable of coding.  And even more importably for its contribution to the overall value of the jparse repo](https://github.com/xexyl/jparse), something you should be able to document so that others may be able to use the JSON semantics facility w/o having to deal with much of the complexity of JSON parse tree data structures.

**Comment by lcn2** [comment]
# JPRINT SUGGESTON

The `jprint -S` as described in [GH-issuecomment-2323573780](https://github.com/xexyl/jparse/issues/5#issuecomment-2323573780) is a **LOWER PRIORTY** than the general `jprint(1)` tool.

We recommend you focus on coding the general `jprint(1)` tool such that it someone invokes `jprint -S ...` the tool simply prints:

> The -S facility has not yet been implemented

and then does an `exit(4)`.

See the command line usage message in the [top level TODO](https://github.com/xexyl/jparse/issues/5#issue-2494233807).

Focus on the general `jprint(1)` tool, leave a command like parsing hook for `-S` and focus on the general tool first.

... post **Great Fork Merge** of course.

**Comment by lcn2** [comment]
# Priority of jval vs JSON semantic checking

We believe that implementing `jval(1)` may be a higher priority than implementing the library functions needed to implement JSON semantic checking.  This is because `jval(1)`, using a facility similar to [JSONPath.sh](https://github.com/lcn2/JSONPath.sh), would be useful to modify JSON documents (see comment regarding "could be parched / edited" in  [GH-issuecomment-2323573780](https://github.com/xexyl/jparse/issues/5#issuecomment-2323573780)).

So the overall priority might be:

1. Work to start the **Great Fork Merge**
2. code `jprint(1)` with `-S` returning an error message and exit 4
3. open the jval(1) issue
4. document the JSON semantic string format
5. code `jprint -S`
6. code jval(1)
7. code the JSON semantic library facility

The above is just a guess at the overall priorities.  Somewhere within that list is to hold the IOCCC28 as well.  :-)

**Comment by lcn2** [comment]
# A mis-feature or BUG in JSONPath.sh

In the other repo, the `bin/jfmt-wrapper.sh` tool uses `JSONPath.sh` (see our [JSONPath.sh repo](https://github.com/lcn2/JSONPath.sh)) to format JSON in a canonical format.

Consider, however, this valid JSON:

```json
{"foo":[100]}
```

When processed the `bin/jfmt-wrapper.sh` tool in the other repo, the tool (via `JSONPath.sh`) produces:

```json
{
    "foo" :     [
        100
    ]
}
```

Notice the excessive whitespace between the ":" and the "[".   That should **NOT** be considered canonical.

# QUESTION

How should this JSON be formatted canonically?

```json
{"foo":[100]}
```

**Comment by xexyl** [comment]
> # A mis-feature or BUG in JSONPath.sh

Are you sure the problem is not JSON itself ? :-)

**Comment by xexyl** [comment]
# QUESTION

What is the order of priority of all the tools (including chkentry mods) versus the other website, please?

**Comment by xexyl** [comment]
# SUGGESTION

For the tools being discussed here why don't you update the JSON_util_README.md (if that is what it was called) to help organise the thoughts? We can then discuss it in more detail here. We might include information here too of course but it might be nice to have a document? Just a thought: maybe not worth it, maybe worth it.

**Comment by xexyl** [comment]
> # QUESTION
> How should this JSON be formatted canonically?
> 
> ```json
> {"foo":[100]}
> ```
> 

Why not have the 4 spaces instead of the tabs and newlines?

```json
{
    "foo" : 
         [
              100
         ]
}
```

or something like that? Maybe the spacing is off but that's one possibility maybe? I have to leave now .. back later today I hope.

> # UPDATE 0
> We ask this because the answer will help is "bug fix / improve" our [JSONPath.sh repo].

Well we can of course discuss the above.
